### DISCUSSION TITLE
How to specify relative font sizes?

### DISCUSSION THREAD

**Discussion by davidchisnall** [sile-typesetter/sile#2182] [Q&A] [answered]
Looking at the code for the verbatim environment, it sets the font size to the current size - 3 (pt?).  I was not able to work out how to write that in the declarative syntax.  The manual mentions that you can write things in terms of the page height or width.

**Comment by Omikhleia** [comment]
> it sets the font size to the current size - 3 (pt?)

It still does? This is likely wrong, there's no reason for using a "hard-coded" 3pt smaller font, not depending on what was the original size. ^^
What you are looking for might be a font size in "em" (e.g. 0.8em = slightly smaller than the current font size)

**Reply by alerque** [reply]
Yes, this does look wrong.

**Reply by davidchisnall** [reply]
This is how I read [the code in verbatim](https://github.com/sile-typesetter/sile/blob/d1ed71e548a858631b083c7da3985f157f0ce809/packages/verbatim/init.lua#L9).

I was trying to achieve the same result as this environment.

**Reply by Omikhleia** [reply]
N.B. I see from some traces of yours in #1837 that you've **resilient.sile** installed, so be aware that when using the "resilient.book" class (instead of the default "book" class), the `\verbatim:font` hook (and actually the whole default `verbatim` package) is not used in principle.
Anyhow, this -3pt correction is a bad default, and the right way to pick a smaller or bigger font than the current font is to use a relative unit, typically in "em". If your initial font, say, is at 10pt, then 1em for it = 10pt, and an environment switching to 0.9em would thus be in 9pt (and so on, the same environment nested would be in 8.1pt)

**Reply by Omikhleia** [reply]
P.S. Just in case old-school typographers pass by, using "em" ratios is a technical answer, and not a "typographical" one in the stricter sense (in the old era of metal typography, a compositor would have frown upon the above 8.1pt of course).

**Reply by Omikhleia** [reply]
> Yes, this does look wrong.

Opened as #2235 = fixed in SILE 0.15.10 see next answer.

**Reply by Omikhleia** [reply, marked as answer]
So now that SILE 0.15.10 is available, we should be able to answer the question fully.

Relative font size can be expressed in a **font-relative unit**.
  - `1em` = the current font size
  - (`1en` = `0.5em`)
  - `1ex` = the height of an "x" in the current font.

The **em** is the most used e.g. for bigger title headings, smaller block quotes, etc.
So if you had, say, a current font at 10pt, and then use `\font[size=0.9em]{smaller}`,  the "smaller text" will now 90% smaller, hence at 9pt.

The **ex** unit expresses a size relative to the x-height and is also sometimes used, mostly for scaling and raising characters in the _same_ font. A poor's man superscript 1 _could_ for instance be `\raise[height=0.7ex]{\font[size=1.5ex]{1}}` -- raised at 70% of a lowercase x, and made smaller _theoretically_ in regular fonts, as the x-height is usually some ratio < 50% of the font size). This is where it gets weird, so let's illustrate with "Libertinus Serif " at 10pt. We can measure: 1em = 10pt (by definition), 1ex = 4.31pt (measured), and hence our "naive superscript here would be raised by approx. 3pt and use a 6.465pt font size.

(For the sake of completeness, other relative units exist, and may or may not be _font_-relative. A paragraph indent or a baseline leading will usually be font-relative, although this is by no mean mandatory.)

**But** as one can start pondering, the em/ex ratio depends on the font design and doesn't really tell how characters fit in that size. _Different_ fonts at the same size (= having the same em) can vary a lot in the size of their letters within that "boundary". In other terms, fonts have different "eyes". Let's illustrate again with the "Hack" font at 10pt, measuring: 1em = 10pt (by definition), 1ex = 5.46875pt (measured)... So when one switches between different fonts, they don't match well side-by-side at the same font size. Here, a lowercase letter in "Hack" is much bigger that a lowercase letter in "Libertinus Serif", and such effects are (most of the time) quite ugly.

Yet, verbatim / code are usually about switching between a regular text font and a (monospace) code font. Unless the text font also comes with a well-made monospace code-looking equivalent (e.g. the Libertinus family has a "Libertinus Mono" font), some **font adjustment** must be made for a better-looking result.

This is now available SILE 0.15.10: `\font[family=Hack, adjust=ex-height]` will compute the target font size so that the "ex" matches that of the current font. (The other possible adjustment is `cap-height`). Let's use this to switch from Libertinus Serif at 10pt, and measure our Hack text: 1ex = 4.31pt (by _construction_ ± rounding), 1em=7.88pt (measured) -- So we are now using a smaller absolute font size for Hack, but the lowercase letters match. It _usually_ looks nicer... It is _always_ a compromise here!

SILE 0.15.10 now uses this technique in `\code` and `\verbatim:font`. For the record, **resilient.sile** 2.8.0 also supports it, though one may have to tweak their _existing_ style files as explained in the [release note](https://github.com/Omikhleia/resilient.sile/releases/tag/v2.8.0).

@davidchisnall Does this answer satisfies your initial question?

**Reply by Omikhleia** [reply]
Just to slightly update and correct my earlier response:
> `1ex`= the height of an "x" in the current font.

_To be really precise,_ this is wrong or somewhat approximate. As discussed elsewhere, the "x" _character_ in a font might have a different height than the intended font-designer's actual "ex" size. 
See #2317 (and related PR) for details and better explanations. But other than that, the above text remains valid (almost, i.e. with slightly different numbers in the examples).

**Reply by Omikhleia** [reply]
@davidchisnall Is there anything else needed for marking this topic as answered?

**Comment by alerque** [comment]
The easiest ways is to use relative *units*. For example the `em` unit is typically the size of the current font, so `\font[size=0.95em]{this is 5% smaller}` would make a smaller size relative to the surrounding context.

Relative sizes are typically not specified as with an absolute value offset from a relative size. You could do that, but you'd probably need to make a small wrapper function of your own to accomplish it:

```lua
SILE.registerCommand("three-smaller", function (options, content)
   options.size = SILE.types.measurement("1em"):absolute() - SILE.types.measurement("3pt")
   SILE.call("font", options, content)
end)
```

Then in your content, you could use `\three-smaller{this is 3 hard points smaller than the surrounding context}`.
